File: plugins/chromecast-display/android-sender/ChromecastCastSkill.kt
package com.smarthome.plugin.chromecast

import com.smarthome.livingroom_android.brain.dto.CapabilityDescriptor
import com.smarthome.livingroom_android.brain.dto.SchemaField
import com.smarthome.livingroom_android.brain.dto.ServiceDescriptor
import com.smarthome.livingroom_android.capability.Capabilities
import com.smarthome.livingroom_android.skill.Skill
import com.smarthome.livingroom_android.skill.SkillContext
import com.smarthome.livingroom_android.skill.SkillResult
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

class ChromecastCastSkill(
    private val onLog: (String) -> Unit = {},
) : Skill {
    override fun service(): ServiceDescriptor =
        ServiceDescriptor(
            serviceId = SKILL_ID,
            version = "0.1.0",
            displayName = "Chromecast Display",
            group = "display",
            capabilities = listOf(
                CapabilityDescriptor(
                    capabilityId = Capabilities.DISPLAY_PHOTO,
                    role = "单图投屏器",
                    plannerRecognize = "把一张已有 Image Asset 投到电视/投屏端。入参 asset_ref（常为 \$asset_ref）。这是计划步，不能只用 presentation.endpoint 代替。多张轮播不要用本步",
                    typicalTriggers = listOf("把这张图投到电视", "投屏", "投到电视", "丢到电视", "放到电视"),
                    doNotDispatch = listOf("拍多图", "幻灯片", "放歌", "按厂商选 Chromecast"),
                    kind = "output",
                    inputSchema = mapOf(
                        "asset_ref" to SchemaField("string", required = true),
                    ),
                ),
            ),
        )

    override suspend fun execute(
        capabilityId: String,
        params: Map<String, Any?>,
        ctx: SkillContext,
    ): SkillResult = withContext(Dispatchers.Main) {
        when (capabilityId) {
            Capabilities.DISPLAY_PHOTO -> {
                val url = params["photo_url"]?.toString()?.trim().orEmpty()
                if (url.isEmpty()) {
                    return@withContext SkillResult.error("missing photo_url")
                }
                val cast = CastSessionController.get(ctx.appContext)
                cast.onLog = onLog
                val result = cast.castPhoto(url)
                if (result.isSuccess) {
                    SkillResult.ok("cast ok → $url")
                } else {
                    SkillResult.error(result.exceptionOrNull()?.message ?: "cast failed")
                }
            }
            else -> SkillResult.error("unsupported capability: $capabilityId")
        }
    }

    companion object {
        const val SKILL_ID = "chromecast.display"
    }
}

File: plugins/chromecast-display/ios/ChromecastCastSkill.swift
import Foundation

/// Wire service chromecast.display — iPhone Cast Sender for display.photo / slideshow.
/// Cast Presentation Protocol V1: docs/chromecast-cast-protocol.md
final class ChromecastCastSkill: Skill {
    static let skillId = "chromecast.display"

    private let cast: CastSessionController

    init(cast: CastSessionController = .shared) {
        self.cast = cast
    }

    func service() -> ServiceDescriptor {
        ServiceDescriptor(
            serviceId: Self.skillId,
            version: "0.5.0",
            displayName: "Chromecast Cast",
            group: "display",
            capabilities: [
                CapabilityDescriptor(
                    capabilityId: "display.photo",
                    description: "能：把本步 asset_ref（AssetRef）经 Cast Presentation Command 投到 Chromecast（F7649303）。仅用户明确要投电视时用。不能：拍照、自己捡图、收永久 URL、TTS。accepted≠已显示；等 Receiver presentation.started。",
                    inputSchema: [
                        "asset_ref": SchemaField(
                            type: "string",
                            required: true,
                            description: "AssetRef JSON {asset_id, type}。禁止 photo_url 作为 identity。"
                        ),
                    ]
                ),
                CapabilityDescriptor(
                    capabilityId: "display.slideshow",
                    description: "能：把本步必填 asset_refs 轮播投到电视（多次 present）。不能：从前序自己拼列表、空数组。",
                    inputSchema: [
                        "asset_refs": SchemaField(
                            type: "string",
                            required: true,
                            description: "AssetRef JSON 数组。"
                        ),
                        "interval_sec": SchemaField(
                            type: "string",
                            required: false,
                            description: "帧间隔秒，默认 5。"
                        ),
                    ]
                ),
            ]
        )
    }

    func execute(capabilityId: String, params: [String: String], context: SkillContext) async -> SkillResult {
        switch capabilityId {
        case "display.photo", Capabilities.displayPhoto:
            // Runtime may hydrate temporary LAN URL into photo_url for Cast transport.
            let url = params["photo_url"]?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
            guard !url.isEmpty else {
                return .error("display.photo requires resolved photo_url (from asset_ref)")
            }
            let assetId = Self.assetId(from: params["asset_ref"])
            let result = await cast.castPhoto(urlString: url, assetId: assetId)
            if result.ok {
                return .ok(result.message)
            }
            return .error(result.message)
        case "display.slideshow":
            let urls = Self.photoURLs(from: params)
            guard !urls.isEmpty else {
                return .error("display.slideshow requires resolved photo URLs from asset_refs")
            }
            let interval = Double(params["interval_sec"] ?? "") ?? 5
            let result = await cast.castSlideshow(urlStrings: urls, intervalSec: interval)
            if result.ok {
                return .ok(result.message)
            }
            return .error(result.message)
        default:
            return .error("unsupported capability: \(capabilityId)")
        }
    }

    private static func assetId(from raw: String?) -> String? {
        guard let raw = raw?.trimmingCharacters(in: .whitespacesAndNewlines), !raw.isEmpty,
              let data = raw.data(using: .utf8),
              let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let id = obj["asset_id"] as? String,
              !id.isEmpty else {
            return nil
        }
        return id
    }

    private static func photoURLs(from params: [String: String]) -> [String] {
        if let raw = params["photo_urls"]?.trimmingCharacters(in: .whitespacesAndNewlines),
           let data = raw.data(using: .utf8),
           let arr = try? JSONSerialization.jsonObject(with: data) as? [Any] {
            return arr.compactMap { ($0 as? String)?.trimmingCharacters(in: .whitespacesAndNewlines) }
                .filter { !$0.isEmpty }
        }
        if let single = params["photo_url"]?.trimmingCharacters(in: .whitespacesAndNewlines), !single.isEmpty {
            return [single]
        }
        return []
    }
}
